### ISSUE TITLE
[Extension]: a1111-sd-webui-haku-img

### ISSUE THREAD

**Issue by HarrisTerry** [vladmandic/automatic#840]
### Issue Description

The main tab to appear at the top does not appear.

Here is the error:
'''15:38:11-473405 ERROR    executing callback: D:\AI\Vlad1111\automatic\extensions\a1111-sd-webui-haku-img\scripts\main.py
                         ui_tabs_callback: UnboundLocalError
╭───────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ D:\AI\Vlad1111\automatic\modules\script_callbacks.py:151 in ui_tabs_callback                                         │
│                                                                                                                      │
│   150 │   │   try:                                                                                                   │
│ ❱ 151 │   │   │   res += c.callback() or []                                                                          │
│   152 │   │   except Exception as e:                                                                                 │
│                                                                                                                      │
│ D:\AI\Vlad1111\automatic\extensions\a1111-sd-webui-haku-img\scripts\main.py:314 in add_tab                           │
│                                                                                                                      │
│   313 │   │   gpc.bind_buttons(send_btns, image_out, None)                                                           │
│ ❱ 314 │   │   for btn, img in zip(btns, all_layers):                                                                 │
│   315 │   │   │   btn.click(lambda x:x, image_out, img)                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
UnboundLocalError: local variable 'btns' referenced before assignment'''

So I am no scripter but I see btns in there :)

This is a very useful extension that allows:
blending (support 17 mode)
color adjust
tone curve
blur
img2sketch(thourgh XDoG)
pixelize(Kmeans or dithering)
glow (blur and blend)
inoutpaint support tool

No rush, but would be great to have in there!

THANK YOU!!

### Version Platform Description

version  5bd863f9 Tue May 9 14:08:57 2023 -0400
Win11, Edge

### URL link of the extension

https://github.com/KohakuBlueleaf/a1111-sd-webui-haku-img

**Comment by Weroxig** [comment]
It would be a good idea to post this in their repository instead of this one.

**Comment by HarrisTerry** [comment]
I can do that, but even if it works in Auto1111 and not this version?

**Comment by BinaryQuantumSoul** [comment]
> I can do that, but even if it works in Auto1111 and not this version?

Try, will depend on the extension developer

**Comment by vladmandic** [comment]
the problem with this extension is that it doesn't check which buttons are present and what they do, it has hard-coded labels as checks. so if i rename any button label, extension fails - that is just bad extension.

for example:
<https://github.com/KohakuBlueleaf/a1111-sd-webui-haku-img/blob/99344e7f5221be81459f5394447ce1d05c6bbefc/scripts/main.py#LL55C1-L56C1>

author should not cater to this repo vs a1111 specifically, just enumerate buttons using some logic, not hard coded label names.

as it is, i'm closing this as will-not-fix on my side as only solution would be to rename all buttons.

**Comment by HarrisTerry** [comment]
Thanks for looking into this.

**Comment by KohakuBlueleaf** [comment]
> the problem with this extension is that it doesn't check which buttons are present and what they do, it has hard-coded labels as checks. so if i rename any button label, extension fails - that is just bad extension.
> 
> for example: https://github.com/KohakuBlueleaf/a1111-sd-webui-haku-img/blob/99344e7f5221be81459f5394447ce1d05c6bbefc/scripts/main.py#LL55C1-L56C1
> 
> author should not cater to this repo vs a1111 specifically, just enumerate buttons using some logic, not hard coded label names.
> 
> as it is, i'm closing this as will-not-fix on my side as only solution would be to rename all buttons.

As I know, If I want to add some extra buttons beside the "send to XXX" btn (or whatever you renamed)
I can only use after_component callback and use the information in it to check which component you created.

The problem is, the information I can get is something like that:
`{'visible': True, 'elem_id': 'extras_tab', 'elem_classes': None, 'value': 'Send to extras', 'interactive': True}`
Everything is hard coded.

Basically I think it is impossible to not use any hard coded information in it.
Since there isn't any hook or api for "adding custom component beside 'send to xxx'". Since "which tab will have send btn" are also hard coded. `buttons = parameters_copypaste.create_buttons(["txt2img", "img2img", "inpaint", "extras"])`

I think I don't have any other method to make my extension "good". But I think the reason is that a1111's original impl is "bad" for ext.
